solr/core/src/test/org/apache/solr/update/processor/UpsertConditionTest.java

```diff
@@ -178,6 +180,87 @@ public void givenNumericField_whenMatching() {
     assertTrue(condition.matches(oldDoc, newDoc));
   }
 
+  @Test
+  public void givenAtomicUpdateSet_whenMatching() {
+    NamedList<String> args = new NamedList<>(ImmutableMap.of(
+        "must", "NEW.field:value",
+        "action", "skip"
+    ));
+
+    UpsertCondition condition = UpsertCondition.parse("skip-it", args);
+
+    assertThat(condition.getName(), is("skip-it"));
+
+    SolrInputDocument oldDoc = new SolrInputDocument();
+    SolrInputDocument newDoc = new SolrInputDocument();
+
+    newDoc.setField("field", Collections.singletonMap("set", "other1"));
+    assertFalse(condition.matches(oldDoc, newDoc));
+
+    newDoc.setField("field", Collections.singletonMap("set", "value"));
+    assertTrue(condition.matches(oldDoc, newDoc));
+
+    newDoc.setField("field", Collections.singletonMap("set", ImmutableList.of("value", "other2")));
+    assertTrue(condition.matches(oldDoc, newDoc));
+
+    newDoc.setField("field", Collections.singletonMap("set", ImmutableList.of("other1", "other2")));
+    assertFalse(condition.matches(oldDoc, newDoc));
+  }
+
+  @Test
+  public void givenAtomicUpdateAdd_whenMatching() {
+    NamedList<String> args = new NamedList<>(ImmutableMap.of(
+        "must", "NEW.field:value",
+        "action", "skip"
+    ));
+
+    UpsertCondition condition = UpsertCondition.parse("skip-it", args);
+
+    assertThat(condition.getName(), is("skip-it"));
+
+    SolrInputDocument oldDoc = new SolrInputDocument();
+    SolrInputDocument newDoc = new SolrInputDocument();
+
+    newDoc.setField("field", Collections.singletonMap("add", "other1"));
+    assertFalse(condition.matches(oldDoc, newDoc));
+
+    newDoc.setField("field", Collections.singletonMap("add", "value"));
+    assertTrue(condition.matches(oldDoc, newDoc));
+
+    newDoc.setField("field", Collections.singletonMap("add", ImmutableList.of("value", "other2")));
+    assertTrue(condition.matches(oldDoc, newDoc));
+
+    newDoc.setField("field", Collections.singletonMap("add", ImmutableList.of("other1", "other2")));
+    assertFalse(condition.matches(oldDoc, newDoc));
+  }
+
+  @Test
+  public void givenAtomicUpdateRemove_whenMatching() {
+    NamedList<String> args = new NamedList<>(ImmutableMap.of(
+        "must", "NEW.field:value",
+        "action", "skip"
+    ));
+
+    UpsertCondition condition = UpsertCondition.parse("skip-it", args);
+
+    assertThat(condition.getName(), is("skip-it"));
+
+    SolrInputDocument oldDoc = new SolrInputDocument();
+    SolrInputDocument newDoc = new SolrInputDocument();
+
+    oldDoc.setField("field", Collections.singletonMap("remove", "other1"));
+    assertFalse(condition.matches(oldDoc, newDoc));
+
+    newDoc.setField("field", Collections.singletonMap("remove", "value"));
+    assertFalse(condition.matches(oldDoc, newDoc));
```

Comment:
Are these false because it's removing the value (hence the resulting new document won't have that value) or is it because it's not supported yet?

Reply:
false because it's not supported.